File: src/image-utils.js
/**
 * Image sizing utilities for aspect-ratio-preserving bounding-box fits.
 * Used by format skills (docx, pptx, pdf) to prevent logo/image squashing.
 */

const sharp = require('../node_modules/sharp');
const path = require('path');

/**
 * Fit an image within a bounding box while preserving aspect ratio.
 * @param {string} imagePath - Path to the image file
 * @param {number} maxWidth - Maximum width (unitless — caller decides units)
 * @param {number} maxHeight - Maximum height (unitless)
 * @returns {Promise<{width: number, height: number, naturalWidth: number, naturalHeight: number}>}
 */
async function fitImageInBox(imagePath, maxWidth, maxHeight) {
  const metadata = await sharp(imagePath).metadata();
  const naturalWidth = metadata.width;
  const naturalHeight = metadata.height;
  const aspectRatio = naturalWidth / naturalHeight;

  let w = maxWidth;
  let h = w / aspectRatio;
  if (h > maxHeight) {
    h = maxHeight;
    w = h * aspectRatio;
  }

  return {
    width: Math.round(w * 100) / 100,
    height: Math.round(h * 100) / 100,
    naturalWidth,
    naturalHeight,
  };
}

/**
 * Parse a charter dimension string like "120pt" into {value, unit}.
 * @param {string} str - e.g. "120pt", "50px", "2in"
 * @returns {{value: number, unit: string}}
 */
function parseDimension(str) {
  const match = String(str).match(/^([\d.]+)\s*([a-z]+)$/i);
  if (!match) throw new Error(`Cannot parse dimension: "${str}"`);
  return { value: parseFloat(match[1]), unit: match[2] };
}

/**
 * Fit a logo within the bounding box defined by a charter's logo config.
 * @param {string} imagePath - Path to the logo image file
 * @param {object} logoConfig - Charter logo object with maxWidth, maxHeight, sizing
 * @param {string} logoConfig.maxWidth - e.g. "120pt"
 * @param {string} logoConfig.maxHeight - e.g. "50pt"
 * @param {string} [logoConfig.sizing] - "contain" (default) or "cover"
 * @returns {Promise<{width: number, height: number, unit: string, naturalWidth: number, naturalHeight: number}>}
 */
async function fitLogoFromCharter(imagePath, logoConfig) {
  const maxW = parseDimension(logoConfig.maxWidth);
  const maxH = parseDimension(logoConfig.maxHeight);

  if (maxW.unit !== maxH.unit) {
    throw new Error(`Mismatched units: maxWidth="${logoConfig.maxWidth}", maxHeight="${logoConfig.maxHeight}"`);
  }

  const result = await fitImageInBox(imagePath, maxW.value, maxH.value);
  return {
    width: result.width,
    height: result.height,
    unit: maxW.unit,
    naturalWidth: result.naturalWidth,
    naturalHeight: result.naturalHeight,
  };
}

module.exports = { fitImageInBox, fitLogoFromCharter, parseDimension };

File: src/image_utils.py
"""Image sizing utilities for aspect-ratio-preserving bounding-box fits.

Used by format skills (docx, pptx, pdf) to prevent logo/image squashing.
"""

from __future__ import annotations

import re
from pathlib import Path

from PIL import Image


def fit_image_in_box(
    image_path: str | Path, max_width: float, max_height: float
) -> dict[str, float]:
    """Fit an image within a bounding box while preserving aspect ratio.

    Args:
        image_path: Path to the image file.
        max_width: Maximum width (unitless — caller decides units).
        max_height: Maximum height (unitless).

    Returns:
        Dict with keys: width, height, natural_width, natural_height.
    """
    with Image.open(image_path) as img:
        natural_width, natural_height = img.size

    aspect_ratio = natural_width / natural_height

    w = max_width
    h = w / aspect_ratio
    if h > max_height:
        h = max_height
        w = h * aspect_ratio

    return {
        "width": round(w, 2),
        "height": round(h, 2),
        "natural_width": natural_width,
        "natural_height": natural_height,
    }


def parse_dimension(s: str) -> tuple[float, str]:
    """Parse a charter dimension string like '120pt' into (value, unit)."""
    match = re.match(r"^([\d.]+)\s*([a-z]+)$", s, re.IGNORECASE)
    if not match:
        raise ValueError(f'Cannot parse dimension: "{s}"')
    return float(match.group(1)), match.group(2)


def fit_logo_from_charter(
    image_path: str | Path, logo_config: dict[str, str]
) -> dict[str, float | str]:
    """Fit a logo within the bounding box defined by a charter's logo config.

    Args:
        image_path: Path to the logo image file.
        logo_config: Charter logo object with maxWidth, maxHeight, sizing.

    Returns:
        Dict with keys: width, height, unit, natural_width, natural_height.
    """
    max_w, unit_w = parse_dimension(logo_config["maxWidth"])
    max_h, unit_h = parse_dimension(logo_config["maxHeight"])

    if unit_w != unit_h:
        raise ValueError(
            f'Mismatched units: maxWidth="{logo_config["maxWidth"]}", '
            f'maxHeight="{logo_config["maxHeight"]}"'
        )

    result = fit_image_in_box(image_path, max_w, max_h)
    return {
        "width": result["width"],
        "height": result["height"],
        "unit": unit_w,
        "natural_width": result["natural_width"],
        "natural_height": result["natural_height"],
    }
